flask/src/main/java/net/corda/gradle/flask/FlaskJarTask.java

```diff
@@ -0,0 +1,240 @@
+import org.gradle.api.Action;
+import org.gradle.api.NamedDomainObjectCollection;
+import org.gradle.api.NamedDomainObjectContainer;
+import org.gradle.api.internal.file.CopyActionProcessingStreamAction;
+import org.gradle.api.internal.file.copy.CopyAction;
+import org.gradle.api.internal.file.copy.CopyActionProcessingStream;
+import org.gradle.api.internal.file.copy.FileCopyDetailsInternal;
+import org.gradle.api.model.ObjectFactory;
+import org.gradle.api.plugins.BasePluginConvention;
+import org.gradle.api.provider.ListProperty;
+import org.gradle.api.provider.Property;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.Nested;
+import org.gradle.api.tasks.WorkResult;
+import org.gradle.api.tasks.bundling.AbstractArchiveTask;
+
+import javax.annotation.Nonnull;
+import javax.inject.Inject;
+import java.io.*;
+import java.security.MessageDigest;
+import java.util.*;
+import java.util.function.Supplier;
+import java.util.jar.Attributes;
+import java.util.jar.JarFile;
+import java.util.jar.Manifest;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
+import java.util.zip.ZipOutputStream;
+
+public class FlaskJarTask extends AbstractArchiveTask {
+
+    private final Property<String> launcherClassName;
+
+    @Input
+    public Property<String> getLauncherClassName() {
+        return launcherClassName;
+    }
+
+    private final Property<String> mainClassName;
+
+    @Input
+    public Property<String> getMainClassName() {
+        return mainClassName;
+    }
+
+    private final ListProperty<String> jvmArgs;
+
+    @Input
+    public ListProperty<String> getJvmArgs() {
+        return jvmArgs;
+    }
+
+    public void includeLibraries(Object files) {
+        into(Flask.Constants.LIBRARIES_FOLDER, (copySpec) -> copySpec.from(files));
+    }
+
+    private final NamedDomainObjectContainer<JavaAgent> javaAgents;
+
+    @Nested
+    public NamedDomainObjectCollection<JavaAgent> getJavaAgents() {
+        return javaAgents;
+    }
+
+    public void javaAgents(@Nonnull Action<NamedDomainObjectCollection<JavaAgent>> action) {
+        action.execute(javaAgents);
+    }
+
+    @Inject
+    public FlaskJarTask(@Nonnull ObjectFactory objects) {
+        BasePluginConvention basePluginConvention = getProject().getConvention().getPlugin(BasePluginConvention.class);
+        getDestinationDirectory().set(basePluginConvention.getLibsDir());
+        getArchiveBaseName().convention(getProject().getName());
+        getArchiveExtension().convention("jar");
+        launcherClassName = objects.property(String.class).convention(Flask.Constants.DEFAULT_LAUNCHER_NAME);
+        mainClassName = objects.property(String.class);
+        jvmArgs = objects.listProperty(String.class);
+        javaAgents = objects.domainObjectContainer(JavaAgent.class);
+        from(getProject().tarTree(LauncherResource.instance), copySpec -> exclude(JarFile.MANIFEST_NAME));
```

Comment:
You will need to remove this line after the changes to `FlaskPlugin` to avoid "duplicate zip entry" exceptions. Gradle will already have included the launcher's files from the `build/classes/flask-launcher` directory.